File: runtime/src/lore/runtime/simplification.ts
import { Function, FunctionType } from './functions.ts'
import { Intersection } from './intersections.ts'
import { Lists, ListType } from './lists.ts'
import { Shape, ShapeType } from './shapes.ts'
import { Sum } from './sums.ts'
import { Tuple, TupleType } from './tuples.ts'
import { DeclaredSchema, DeclaredSchemas } from './types/declared-schemas.ts'
import { DeclaredType, InvalidTypeArguments } from './types/declared-types.ts'
import { Kind } from './types/kinds.ts'
import { isSubtype } from './types/subtyping.ts'
import { Variance } from './types/type-variables.ts'
import { Type } from './types/types.ts'
import { allExcluding, appendInner, flattenedUnique } from './types/util.ts'

/**
 * See the corresponding compiler implementation `lore.compiler.types.Simplification`.
 */
export const Simplification = {
  construct(kind: Kind, parts: Array<Type>): Type {
    if (parts.length === 1) {
      return parts[0]
    }

    const flattened = flattenedUnique(kind, parts)
    const combined = combine(flattened, kind)
    const relevants = filterRelevantTypes(combined, kind)

    if (relevants.length == 1) {
      return relevants[0]
    } else {
      switch (kind) {
        case Kind.Sum: return Sum.type(relevants)
        case Kind.Intersection: return Intersection.type(relevants)
        default: throw invalidKind(kind)
      }
    }
  },
}

function combine(parts: Array<Type>, kind: Kind): Array<Type> {
  let constructCovariant: (parts: Array<Type>) => Type
  let constructContravariant: (parts: Array<Type>) => Type

  switch (kind) {
    case Kind.Sum:
      constructCovariant = Sum.simplified
      constructContravariant = Intersection.simplified
      break

    case Kind.Intersection:
      constructCovariant = Intersection.simplified
      constructContravariant = Sum.simplified
      break

    default: throw invalidKind(kind)
  }

  const tuplesBySize: Map<number, Array<TupleType>> = new Map()
  const functions: Array<FunctionType> = []
  const lists: Array<ListType> = []
  const shapes: Array<ShapeType> = []
  const declaredTypesBySchema: Map<DeclaredSchema, Array<DeclaredType>> = new Map()
  const result: Array<Type> = []

  for (const part of parts) {
    switch (part.kind) {
      case Kind.Tuple:
        let tuple = <TupleType> part
        appendInner(tuplesBySize, tuple.types.length, tuple)
        break

      case Kind.Function:
        functions.push(<FunctionType> part)
        break

      case Kind.List:
        lists.push(<ListType> part)
        break

      case Kind.Shape:
        if (kind == Kind.Intersection) {
          shapes.push(<ShapeType> part)
        } else {
          result.push(part)
        }
        break

      case Kind.Trait:
      case Kind.Struct:
        let dt = <DeclaredType> part
        if (!dt.schema.hasInvariantParameters) {
          appendInner(declaredTypesBySchema, dt.schema, dt)
        } else {
          result.push(part)
        }
        break

      default:
        result.push(part)
    }
  }

  // (A, B) | (C, D) :=: (A | C, B | D)
  // (A, B) & (C, D) :=: (A & C, B & D)
  for (const [size, tuples] of tuplesBySize) {
    if (tuples.length === 1) {
      result.push(tuples[0])
    } else {
      const elements: Array<Type> = []
      for (let i = 0; i <= size; i += 1) {
        const elementParts = []
        for (const tuple of tuples) {
          elementParts.push(tuple.types[i])
        }
        elements.push(constructCovariant(elementParts))
      }
      result.push(Tuple.type(elements))
    }
  }

  // (A => B) | (C => D) :=: (A & C) => (B | D)
  // (A => B) & (C => D) :=: (A | C) => (B & D)
  if (functions.length > 0) {
    const input = Tuple.tupled(constructContravariant(functions.map(f => f.input)))
    const output = constructCovariant(functions.map(f => f.output))
    result.push(Function.type(input, output))
  }

  // [A] | [B] :=: [A | B]
  // [A] & [B] :=: [A & B]
  if (lists.length > 0) {
    const element = constructCovariant(lists.map(t => t.element))
    result.push(Lists.type(element))
  }

  // { name: A } & { name: B } & { health: Int }  ==  { name: A & B, health: Int }
  if (shapes.length > 0) {
    result.push(Shape.combine(shapes))
  }

  // Given a declared type `D[+X, -Y]`:
  // D[A, B] | D[C, D] :=: D[A | C, B & D]
  // D[A, B] & D[C, D] :=: D[A & C, B | D]
  for (const [schema, dts] of declaredTypesBySchema) {
    if (dts.length === 1) {
      result.push(dts[0])
    } else {
      const parameters = schema.typeParameters
      const typeArguments: Array<Type> = []

      for (let i = 0; i < parameters.length; i += 1) {
        const parameter = parameters[i]
        // Note that `typeArguments` will definitely be defined, because the schema has at least one parameter.
        const argumentParts = dts.map(dt => (<Array<Type>> dt.typeArguments)[i])
        switch (parameter.variance) {
          case Variance.Covariant:
            typeArguments.push(constructCovariant(argumentParts))
            break

          case Variance.Contravariant:
            typeArguments.push(constructContravariant(argumentParts))
            break

          case Variance.Invariant:
            throw Error(`At this stage, a type parameter may not be invariant. Type schema: ${schema.name}.`)
        }
      }

      // Note that this instantiation ignores a struct type's open property types. This is okay, as combined types
      // shouldn't contain open property types anyway.
      try {
        result.push(DeclaredSchemas.type(schema, typeArguments))
      } catch (e) {
        if (e instanceof InvalidTypeArguments) {
          // If the schema cannot be instantiated with the combined type arguments, default to the uncombined declared
          // types.
          result.push(...dts)
        } else {
          throw e
        }
      }
    }
  }

  return result
}

function filterRelevantTypes(parts: Array<Type>, kind: Kind): Array<Type> {
  switch (kind) {
    case Kind.Sum: return allExcluding(parts, (self, other) => isSubtype(self, other))
    case Kind.Intersection: return allExcluding(parts, (self, other) => isSubtype(other, self))
    default: throw invalidKind(kind)
  }
}

function invalidKind(kind: Kind): Error {
  return Error(`Cannot construct a sum or intersection type from kind ${kind}.`)
}

File: compiler/src/lore/compiler/types/Simplification.scala
package lore.compiler.types

import lore.compiler.core.CompilationException
import lore.compiler.types.TypeVariable.Variance
import lore.compiler.utils.CollectionExtensions.VectorMapExtension

object Simplification {

  /**
    * Constructs a type of `kind`, either a sum type or intersection type, from the given parts. This includes
    * flattening, combining types according to their variance, and taking only the most general/specific types.
    *
    * If a subterm is covariant, the type candidates are combined with the constructor as `tpe` (sum or intersection).
    * If a subterm is contravariant, the opposite constructor is chosen (intersection or sum). Invariant subterms
    * cannot be combined, meaning types with invariant subterms are left as is.
    *
    * The following kinds of types will be simplified: Tuples of the same size, functions, lists, shapes (intersection
    * types only), and declared types of the same schema.
    *
    * TODO (maps): Before this can be implemented for maps, we have to clear up the variance of maps.
    */
  def construct(kind: Kind, parts: Vector[Type]): Type = {
    val flattened = flatten(parts, kind)
    val combined = combine(flattened, kind)
    val relevants = filterRelevantTypes(combined, kind)

    if (relevants.size == 1) {
      relevants.head
    } else {
      kind match {
        case Kind.Sum => SumType(relevants)
        case Kind.Intersection => IntersectionType(relevants)
        case _ => invalidKind(kind)
      }
    }
  }

  private def flatten(parts: Vector[Type], kind: Kind): Vector[Type] = {
    parts.flatMap {
      case SumType(parts) if kind == Kind.Sum => parts
      case IntersectionType(parts) if kind == Kind.Intersection => parts
      case t => Vector(t)
    }
  }

  private def combine(parts: Vector[Type], kind: Kind): Vector[Type] = {
    val constructSum = (parts: Vector[Type]) => SumType.construct(parts)
    val constructIntersection = (parts: Vector[Type]) => IntersectionType.construct(parts)
    val (constructCovariant, constructContravariant) = kind match {
      case Kind.Sum => (constructSum, constructIntersection)
      case Kind.Intersection => (constructIntersection, constructSum)
      case _ => invalidKind(kind)
    }

    var tuplesBySize = Map.empty[Int, Vector[TupleType]]
    var functions = Vector.empty[FunctionType]
    var lists = Vector.empty[ListType]
    var shapes = Vector.empty[ShapeType]
    var declaredTypesBySchema = Map.empty[DeclaredSchema, Vector[DeclaredType]]
    var result = Vector.empty[Type]

    parts.foreach {
      case tuple@TupleType(elements) => tuplesBySize = tuplesBySize.appended(elements.length, tuple)
      case function: FunctionType => functions = functions :+ function
      case list: ListType => lists = lists :+ list
      case shape: ShapeType if kind == Kind.Intersection => shapes = shapes :+ shape
      case dt: DeclaredType if !dt.schema.hasInvariantTypeParameters => declaredTypesBySchema = declaredTypesBySchema.appended(dt.schema, dt)
      case t => result = result :+ t
    }

    // (A, B) | (C, D) :=: (A | C, B | D)
    // (A, B) & (C, D) :=: (A & C, B & D)
    tuplesBySize.foreach {
      case (_, Vector(tuple)) => result = result :+ tuple
      case (_, tuples) => result = result :+ TupleType(tuples.map(_.elements).transpose.map(constructCovariant))
    }

    // (A => B) | (C => D) :=: (A & C) => (B | D)
    // (A => B) & (C => D) :=: (A | C) => (B & D)
    if (functions.nonEmpty) {
      result = result :+ FunctionType(
        Type.tupled(constructContravariant(functions.map(_.input))),
        constructCovariant(functions.map(_.output))
      )
    }

    // [A] | [B] :=: [A | B]
    // [A] & [B] :=: [A & B]
    if (lists.nonEmpty) {
      result = result :+ ListType(constructCovariant(lists.map(_.element)))
    }

    // Note that shape type combining is only performed for intersection types, but this has already been checked
    // during categorization. Hence, `shapes` is only non-empty if `tpe` is an intersection type.
    // { name: A } & { name: B } & { health: Int }  ==  { name: A & B, health: Int }
    if (shapes.nonEmpty) {
      result = result :+ ShapeType.combine(shapes)
    }

    // Given a declared type `D[+X, -Y]`:
    // D[A, B] | D[C, D] :=: D[A | C, B & D]
    // D[A, B] & D[C, D] :=: D[A & C, B | D]
    declaredTypesBySchema.foreach {
      case (_, Vector(dt)) => result = result :+ dt
      case (schema, declaredTypes) =>
        val allArguments = declaredTypes.map(_.typeArguments).transpose
        val combinedArguments = schema.typeParameters.zip(allArguments).map {
          case (parameter, arguments) => parameter.variance match {
            // We've already established that the declared type has no invariant parameters during categorization.
            case Variance.Covariant => constructCovariant(arguments)
            case Variance.Contravariant => constructContravariant(arguments)
            case _ => throw CompilationException(s"At this stage, a type parameter may not be invariant. Type schema: $schema.")
          }
        }
        schema.instantiate(combinedArguments) match {
          case Some(dt) => result = result :+ dt
          case None =>
            // If the schema cannot be instantiated with the combined arguments, for example because lower or upper
            // bounds are violated, we default to the uncombined declared types.
            result = result ++ declaredTypes
        }
    }

    result
  }

  private def filterRelevantTypes(parts: Vector[Type], kind: Kind): Set[Type] = {
    kind match {
      case Kind.Sum => parts.filterNot(t => parts.exists(t < _)).toSet
      case Kind.Intersection => parts.filterNot(t => parts.exists(_ < t)).toSet
      case _ => invalidKind(kind)
    }
  }

  private def invalidKind(kind: Kind): Nothing = {
    throw CompilationException(s"Cannot construct a sum or intersection type from kind $kind.")
  }

}
